### DISCUSSION TITLE
JBang CLI with Rust?

### DISCUSSION THREAD

**Discussion by linux-china** [jbangdev/jbang#1668] [Ideas]
Now JBang CLI written by shell script, any idea to write with Rust?  

Pros:

- Portable 
- Faster than script
- Easy to add some features

**Comment by maxandersen** [comment]
I'm assuming you mean the small wrapper script not jbang cli itself which is mainly java to get access to maven etc.? 

May I ask what you mean by "easy to add some features" ? 

Would it not be simpler to remove the shell script completely and just become graalvm possibly quarkus based cli?

**Reply by linux-china** [reply]
@maxandersen  "easy to add some features" means that some JBang features could be written by Rust: 

-  dynamic shell completion for jbang-catalog.json:  `jbang run |` to add alias name completion.  Some tools, such as [just](https://github.com/casey/just) uses `just --summary` for dynamic shell completion.  Java is not good for this case because of slow startup.
-  some features, such as jdk, trust, template management, could be written by Rust, and it will be fast to invoke. 
-  JSON dump:  dump all JBang information to json, such as jdk, template, config, alias etc, and could be used by other command line or script.  I want to get some information from JBang in the script, and now JBang still slow to dump this information.

I'm not sure quarkus cli with GraalVM native image good or not, but Rust binary is really smaller and faster than Graal native image binary.

**Reply by maxandersen** [reply]
Completion of aliases etc.  would require Almost all of jbang to be written or duplicated in rust.

That ain't in the cards. 

Java with graalvm/quarkus starts in milliseconds - not sure how much faster you think rust can get but I don't currently believe it's fast enough to justify a full rewrite. 

The slow part we have currently is picocli. That's what we should fix.
